### PR TITLE
Add functions to calculate running moments of time series in new time_series sub module

### PR DESCRIPTION
**Pull request by gregordecristoforo**
This PR adds a time_series sub module with two functions for calculating running moments of time series. If this contribution gets approved I will add additional functionality for time series analysis in later PRs (as discussed with @namurphy during the Plasmapy office hours).

Looking forward to your feedback!

### CODE REVIEW

**Comment by StanczakDominik** [comment]
I don't have context from your discussion with @namurphy, so I'll leave questions of package scope for later (Nick, I assume you've thought of that? :) ).

Instead, since this is pure numerics and as "packaged" code that lots of people would (hopefully! :D) use, I think we really need to focus on the correctness and efficiency aspects.

In this context, for the mean, I'll point out the existence of https://stackoverflow.com/a/43200476 -the one with uniform filter. It seems to handle boundaries, cumulative floating point errors better while being faster. I'm not yet sure what the best way to tackle running variance etc would be, I'll look at them later.

There's also pandas' running mean, though that requires a dataframe load. But that's not a big ask. Need to think about just using that, too.

**Comment by gregordecristoforo** [comment]
Thanks for your comment. I think I should provide a bit more context to this PR (sorry for not being explicit earlier, I should have though about the fact that @namurphy isn't the only maintainer of Plasmapy):

My group at UiT has specialized in analyzing and stochastic modelling of intermittent time series in the boundary of fusion plasmas. For this purpose we have developed some small python packages that we use internally and are available [here](https://github.com/uit-cosmo). My idea was to port some of our routines of [our time series analysis repo](https://github.com/uit-cosmo/fpp-analysis-tools) to Plasmapy in order to make the accessible to a bigger user base. The repo contains some useful routines for conditional averaging, calculating correlation functions, excess statistics, deconvolution methods etc.

File: plasmapy/analysis/time_series/running_moments.py
Comment (namurphy) [inline plasmapy/analysis/time_series/running_moments.py:59]: After adding the `api_static` pages, it would be worth checking the Read the Docs documentation preview to make sure this is showing up the way that you intended in the online documentation.

File: plasmapy/analysis/time_series/running_moments.py
Comment (namurphy) [inline plasmapy/analysis/time_series/running_moments.py:99]:
```suggestion
    if time is not None and (len(signal) != len(time)):
```

File: plasmapy/analysis/time_series/running_moments.py
Comment (namurphy) [inline plasmapy/analysis/time_series/running_moments.py:128]: Would it make sense to return a [`collections.namedtuple`](https://docs.python.org/3/library/collections.html#collections.namedtuple) here with `run_moment` and `time` as its attributes?  It would still be possible to do `run_moment, time = running_mean(...)` but it would then also be possible to do `result = running_mean(...)` and then use `result.run_moment` and `result.time`.  I find this helpful when a function returns multiple values that are related to each other.

File: plasmapy/analysis/time_series/running_moments.py
Comment (namurphy) [inline plasmapy/analysis/time_series/running_moments.py:1]: It's great to have this added!

File: plasmapy/analysis/time_series/running_moments.py
Comment (namurphy) [inline plasmapy/analysis/time_series/running_moments.py:20]:
```suggestion
    Calculate the running mean of a sequence.

    Parameters
    ----------
    signal : 1D |array_like|
        Signal to be averaged.

    radius : int
        The window size is ``2 * radius + 1``.

    Returns
    -------
    |array_like|
        Running mean of signal with length ``len(signal) - 2 * radius``.
```
Only for matching formatting.  The double back ticks will format it as literal text that doesn't get linked.

File: plasmapy/analysis/time_series/running_moments.py
Comment (namurphy) [inline plasmapy/analysis/time_series/running_moments.py:62]:
```suggestion
    time : |array_like|, optional
        Time base of ``signal``.
```

File: plasmapy/analysis/time_series/running_moments.py
Comment (namurphy) [inline plasmapy/analysis/time_series/running_moments.py:84]:
```suggestion
    `ValueError`
        If ``moment`` is not in (1, 2, 3, 4).

    `ValueError`
        If ``signal`` and ``time`` don't have same length.

    `ValueError`
        If ``len(signal) <= 4 * radius`` and ``moment > 1``.
```

File: plasmapy/analysis/time_series/running_moments.py
Comment (namurphy) [inline plasmapy/analysis/time_series/running_moments.py:125]:
```suggestion
    if time is not None:
```

**Review by namurphy — CHANGES_REQUESTED** [review]
Thank you, @gregordecristoforo!  I'm excited to see `plasmapy.analysis.time_series` get added.  Most of my suggestions are pretty minor and primarily for consistency with other docstrings in PlasmaPy. 

I'm wondering if it would be helpful to have a check to make sure that `radius` is an integer (more specifically, a subclass of `numbers.Integral`), and to raise a `TypeError` otherwise.  It would add clutter to the code, but would provide a more useful error message for end users. I'll leave it to you to decide.

Since this pull request is adding a new module, we'll need to do one extra step and [create `api_static` files](https://docs.plasmapy.org/en/stable/contributing/doc_guide.html#narrative-documentation-guidelines) so that the subpackage and module get included in the online documentation.

**Comment by namurphy** [comment]
> I don't have context from your discussion with @namurphy, so I'll leave questions of package scope for later (Nick, I assume you've thought of that? :) ).

We had talked about the possibility that the time series analysis functionality could go into a package like SciPy, in particular since a lot of it would be useful in a wide range of fields.  The main reasons why I was thinking it would be appropriate for PlasmaPy are that time series show up in plasma physics a lot, some of the functionality is likely to be somewhat specific to plasma physics, and we want to make sure that there's interoperability with `Quantity` objects and future plasma objects & data sets. This would also be functionality that could be used to link PlasmaPy with other Python in Heliophysics Community packages that work with in situ space plasma data. ☀️ 🛰️ 🌃 

On that note, I've also been wondering is if some of these time series analysis tools could be applied to multi-dimensional data sets too, like solar observations.  That's probably out-of-scope for this pull request though!

Thank you again!

File: plasmapy/analysis/time_series/running_moments.py
Comment (gregordecristoforo) [reply plasmapy/analysis/time_series/running_moments.py:69]: Yes!

File: plasmapy/analysis/time_series/running_moments.py
Comment (gregordecristoforo) [reply plasmapy/analysis/time_series/running_moments.py:128]: Good idea. This is implemented in the updated code.

**Comment by gregordecristoforo** [comment]
Thank you for your suggestions @namurphy!

> I'm wondering if it would be helpful to have a check to make sure that radius is an integer (more specifically, a subclass of numbers.Integral), and to raise a TypeError otherwise. 

I added the check since the `running_mean` function remains relatively short.

>In this context, for the mean, I'll point out the existence of https://stackoverflow.com/a/43200476 -the one with uniform filter. It seems to handle boundaries, cumulative floating point errors better while being faster. I'm not yet sure what the best way to tackle running variance etc would be, I'll look at them later.

This is an interesting Idea. Replacing the numpy implementation of running_mean with scipy.ndimage should be straight forward. The running_moment function shouldn't be affected except of the borders. I will take a look at this.

> On that note, I've also been wondering is if some of these time series analysis tools could be applied to multi-dimensional data sets too, like solar observations. That's probably out-of-scope for this pull request though!

What is the exact form of these data sets? Are you using xarray or something similar? Are you planning to open a separate issue for this or would you like this to be addressed in this PR?

There seems to be an Issue with building the documentation for plasmapy.analysis.time_series after I created a api_staic file. May I ask what is missing?

Thanks again for your feedback @namurphy and @StanczakDominik !

**Comment by namurphy** [comment]
> > On that note, I've also been wondering is if some of these time series analysis tools could be applied to multi-dimensional data sets too, like solar observations. That's probably out-of-scope for this pull request though!
> 
> What is the exact form of these data sets? Are you using xarray or something similar? Are you planning to open a separate issue for this or would you like this to be addressed in this PR?

This could range from `astropy.units.Quantity` objects to `sunpy.map.Map` objects, in addition to NumPy arrays.  We probably shouldn't worry about that in this PR, and it might be best to wait until we have a feature request or specific use case for it before actually going about and implementing it.
 
> Thanks again for your feedback @namurphy and @StanczakDominik !

Thank you for submitting this contribution!

File: plasmapy/analysis/time_series/running_moments.py
Comment (namurphy) [reply plasmapy/analysis/time_series/running_moments.py:74]: It looks like the formatting in the [doc preview](https://plasmapy--1803.org.readthedocs.build/en/1803/api/plasmapy.analysis.time_series.running_moments.running_moment.html#running-moment) is coming out a little weird.  I can never remember the formatting for lists in parameter descriptions, but there are a few examples in [`thomson.py`](https://github.com/PlasmaPy/PlasmaPy/blob/17aecc2aa9dcfbaa03252230f51ab0e229cb7759/plasmapy/diagnostics/thomson.py#L688) for unordered lists.

**Comment by namurphy** [comment]
> I am struggling to understand the reoccurring `InvocationError` when building the docs. On my local machine this error appears when running `tox -e build_docs -- -q` once but the docs are build successfully when running the same command again. May I ask for your opinion on this issue @namurphy?

One of the reasons for recent doc build failures has been the `6.18.*` releases of ipykernel.  If you do `pip install ipykernel==6.17`, then that should resolve that particular issue.  

There's still the remaining `reference target not found` error.  Could you try adding an `api_static` stub file for `plasmapy.analysis.time_series` to see if that goes away?  I'm not certain that will fix it, but that's my best guess.  If that does fix it, I should probably add that to #1817.  

> Slightly out of topic: Once this PR is finally merged I will implement the `excess_statistics` and `conditional_averaging` modules from [this repo](https://github.com/uit-cosmo/fpp-analysis-tools). Creating a notebook in `docs/notebooks/analysis/time_series/` would also be a good idea once enough time series functionality is implemented.

Awesome!  You have our thanks!

**Comment by gregordecristoforo** [comment]
> There's still the remaining reference target not found error. Could you try adding an api_static stub file for plasmapy.analysis.time_series to see if that goes away? I'm not certain that will fix it, but that's my best guess. If that does fix it, I should probably add that to https://github.com/PlasmaPy/PlasmaPy/pull/1817.

Thanks, this has resolved the `reference target not found` error. 

Glad all checks finally pass. Is there anything else left to do on my side?

**Review by namurphy — COMMENTED** [review]
Here are a few final suggestions, most of which are pretty minor.  It's looking really close, and I think we can get it merged pretty soon. Thank you again for submitting this!

File: plasmapy/analysis/time_series/running_moments.py
Comment (gregordecristoforo) [reply plasmapy/analysis/time_series/running_moments.py:139]: I chose `elif` for `moment == 3` and `4`. I think this makes the code more readable.

File: plasmapy/analysis/time_series/running_moments.py
Comment (gregordecristoforo) [reply plasmapy/analysis/time_series/running_moments.py]: Makes sense, I adopted `difference`.

File: plasmapy/analysis/time_series/running_moments.py
Comment (gregordecristoforo) [reply plasmapy/analysis/time_series/running_moments.py]: Renaming it to be private makes the most sense to me.

**Review by namurphy — APPROVED** [review]
LGTM!  I'll set it to auto-merge once the tests pass.  And thank you again for doing this!

**Comment by namurphy** [comment]
Looks like there was a Python 3.10 error that I've run into before, I think related to #750. I'll have to look into this later today.

**Comment by namurphy** [comment]
It looks like the changes from #1822 will need to be merged in before the tests can finish running.  Could you do a `git pull upstream main` (where `upstream` is the git remote for PlasmaPy's central repository) and then push the changes to this branch?  So long as the tests pass, I think this is ready to merge.  Thank you again!

**Comment by gregordecristoforo** [comment]
Sorry to bother you again, but I would really like to understand this issue. 

It seems like the [code coverage ](https://app.codecov.io/gh/PlasmaPy/PlasmaPy/commit/76dc905b36a5d0632ea2b1c915f9f62d9876633b)fails in `plasmapy/__init__.py`. How can that be since I haven't changed this file in this PR?

**Comment by namurphy** [comment]
> It seems like the [code coverage ](https://app.codecov.io/gh/PlasmaPy/PlasmaPy/commit/76dc905b36a5d0632ea2b1c915f9f62d9876633b)fails in `plasmapy/__init__.py`. How can that be since I haven't changed this file in this PR?

The code coverage metric sometimes acts a little weird after doing a `git pull upstream main`.  I'm not sure exactly why, but I suspect it's because it's using the number of lines from when the PR branch diverged from `main` instead of the current state of `main`.  In any case, the ❌ for `codecov/project` doesn't have any meaningful relation to this PR, so I'm happy to merge this now.  Sorry about the delay with the issue with tests!  And you have our sincere thanks for adding this in!

**Pull request merged**
